### PR TITLE
Maintenance

### PR DESCRIPTION
**Pull request by Emadgnia**
New way of getting maintenance info from Firebase instead

### CODE REVIEW

File: Tests/HelpersTests/HelpersTests.swift [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        let expectedOutput = ComparisonResult.orderedAscending
+        XCTAssertEqual(input, expectedOutput, "The output is not expected!")
+    }
+    func testVersionCompareOrderedAscending1Version() {
+        let versionOne = "1"
+        let versionTwo = "2.0.0"
+        let input = versionOne.versionCompare(versionTwo)
+        let expectedOutput = ComparisonResult.orderedAscending
+        XCTAssertEqual(input, expectedOutput, "The output is not expected!")
+    }
```
Comment (patrickusiewicz) [inline Tests/HelpersTests/HelpersTests.swift:32]: Why did you move these test out of `VersionCheckTest`?

File: Tests/MaintenanceTests/MaintenanceTests.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    func testBannerDetails() async throws {
+        let input = Maintenance.isThereAnyMaintenance(
+            fullscreen: false,
+            banner: true,
+            fullscreenDetail: Maintenance.Detail(title: "FullScreen", body: "Body Test"),
+            bannerDetail: Maintenance.Detail(title: "Banner", body: "Body Test"))
+        let expectedOutput = Maintenance.Detail(title: "Banner", body: "Body Test")
+        XCTAssertEqual(input?.1, expectedOutput, "The output is not expected!")
+    }
```
Comment (patrickusiewicz) [inline Tests/MaintenanceTests/MaintenanceTests.swift:49]: An idea: Couldn't you merge these test with `testFullscreen` and `testBanner`? By merging them you would have less maintenance, because the input data for the tests are the same.

File: Tests/MaintenanceTests/MaintenanceTests.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+//  MaintenanceTests.swift
+//
+//
+//  Created by Emad Ghorbania on 18/11/2022.
+//
+
+import XCTest
+@testable import Maintenance
+import Firebase
+final class MaintenanceTests: XCTestCase {
```
Comment (patrickusiewicz) [inline Tests/MaintenanceTests/MaintenanceTests.swift:11]: Should we also have a test case for both `fullscreen` and `banner` are false?

File: Sources/Maintenance/Maintenance.swift [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+//
+//  Created by Emad Ghorbania on 14/12/2022.
+//
+
+import Foundation
+import FirebaseRemoteConfig
+import Helpers
+
+struct Maintenance {
+    var fetch: (_ minimumFetchInterval: Double?,
```
Comment (patrickusiewicz) [inline Sources/Maintenance/Maintenance.swift:13]: Can you point me to where `fetch` is used?

File: Sources/Maintenance/Maintenance.swift [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                _ bannerKey: String,
+                _ fullscreenKey: String,
+                _ bannerDetailsKey: String,
+                _ fullscreenDetailsKey: String) async throws -> (StatusType, Detail)?
+    
+    public static let live = Self { minimumFetchInterval, bannerKey, fullscreenKey, bannerDetailsKey, fullscreenDetailsKey in
```
Comment (patrickusiewicz) [inline Sources/Maintenance/Maintenance.swift:19]: Shouldn't both have DocC comments?

File: Sources/Maintenance/Maintenance.swift [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    ///     }}
+    /// } catch {
+    ///     Send error to crashlytics
+    /// }
+    ///````
+    /// - Parameters:
+    ///   - bannerKey: The Firebase Remote Config key for banner
+    ///   - fullscreenKey: The Firebase Remote Config key for fullscreen
+    ///   - bannerDetailsKey: The Firebase Remote Config key for bannerDetails
+    ///   - fullscreenDetailsKey: The Firebase Remote Config key for fullscreenDetails
```
Comment (patrickusiewicz) [inline Sources/Maintenance/Maintenance.swift:52]: From the description Im not sure what value does these keys point at.
Are the keys pointing a string values that can be shown in the app?

File: Sources/Maintenance/Maintenance.swift [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        
+        
+    }
+    
+    static func isThereAnyMaintenance(fullscreen: Bool,
+                                      banner: Bool,
+                                      fullscreenDetail: Detail,
+                                      bannerDetail: Detail) -> (StatusType, Detail)? {
+        
+        if fullscreen {
```
Comment (patrickusiewicz) [inline Sources/Maintenance/Maintenance.swift:111]: Would it make sense to log if both `fullscreen` & `banner` are true? Then the client of this library knows if something is wrongly configured in Remote Config and can react on it.

File: Sources/Helpers/RemoteConfigHelper.swift [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    public static func getModeledValueForKey<DataType: Codable>(remoteConfig: RemoteConfig,
+                                                         key: String,
+                                                         expectationModel: DataType.Type) throws -> DataType {
+        let data = remoteConfig.configValue(forKey: "\(key)").dataValue
+        return try JSONDecoder().decode(DataType.self, from: data)
+    }
+
+    public static func getBoolValueForKey(remoteConfig: RemoteConfig, key: String) -> Bool {
+        return remoteConfig.configValue(forKey: "\(key)").boolValue
```
Comment (patrickusiewicz) [inline Sources/Helpers/RemoteConfigHelper.swift:48]: Missing `else { throw CustomError.noValueFound(key: key)}`?

File: Sources/Helpers/RemoteConfigHelper.swift [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+                } else {
+                    continuation.resume(returning: .success(remoteConfig))
+                }
+            }
+        }
+    }
+    public static func getModeledValueForKey<DataType: Codable>(remoteConfig: RemoteConfig,
+                                                         key: String,
+                                                         expectationModel: DataType.Type) throws -> DataType {
+        let data = remoteConfig.configValue(forKey: "\(key)").dataValue
```
Comment (patrickusiewicz) [inline Sources/Helpers/RemoteConfigHelper.swift:43]: Missing `else { throw CustomError.noValueFound(key: key)}`?

File: Sources/Helpers/CustomError.swift [hunk 10]
```diff
@@ -0,0 +1,13 @@
+//
+//  File.swift
+//  
+//
+//  Created by Emad Ghorbania on 14/12/2022.
+//
+
+import Foundation
+public enum CustomError: Error, LocalizedError, Equatable {
```
Comment (patrickusiewicz) [inline Sources/Helpers/CustomError.swift:9]: Should we have DocC comments here?

File: Sources/Helpers/RemoteConfigHelper.swift [hunk 11]
```diff
@@ -0,0 +1,65 @@
+//
+//  File.swift
+//  
+//
+//  Created by Emad Ghorbania on 11/11/2022.
+//
+
+import FirebaseRemoteConfig
+
+public struct RemoteConfigHelper {
```
Comment (patrickusiewicz) [inline Sources/Helpers/RemoteConfigHelper.swift:10]: Should we have DocC comments here?

Comment (patrickusiewicz) [reply Sources/Helpers/RemoteConfigHelper.swift:10]: Is `RemoteConfigHelper` only intended to be used internally in this library or is it OK that it is exposed to clients of this library?

File: Sources/Maintenance/Models/Detail.swift [hunk 12]
```diff
@@ -0,0 +1,14 @@
+//
+//  File.swift
+//  
+//
+//  Created by Emad Ghorbania on 14/12/2022.
+//
+
+import Foundation
+extension Maintenance {
```
Comment (patrickusiewicz) [inline Sources/Maintenance/Models/Detail.swift:9]: Should we have DocC comments here?

File: Sources/Maintenance/Models/StatusType.swift [hunk 13]
```diff
@@ -0,0 +1,14 @@
+//
+//  File.swift
+//  
+//
+//  Created by Emad Ghorbania on 14/12/2022.
+//
+
+import Foundation
+extension Maintenance {
```
Comment (patrickusiewicz) [inline Sources/Maintenance/Models/StatusType.swift:9]: Should we have DocC comments here?

File: Sources/VersionCheck/VersionCheck.swift [hunk 14]
```diff
@@ -6,10 +6,21 @@
 //
 
 import FirebaseRemoteConfig
-
+import Helpers
 
 public struct VersionCheck {
-    fileprivate var remoteConfig: RemoteConfig
+    var fetch: (_ latestVersionKeyFromFirebase: String,
```
Comment (patrickusiewicz) [inline Sources/VersionCheck/VersionCheck.swift:12]: Same question as in `Maintenance.swift`.

**Review by patrickusiewicz — COMMENTED** [review]
The reason I haven't accepted the PR is because to me it looks like we are missing couple of checks like `else { throw CustomError.noValueFound(key: key)}` in `RemoteConfigHelper`.

Besides that I have added some comments where to me it looks like we are missing DocC comments. Let me know if Im wrong 🙂 But nothing major.

I dont know if you have planned it, but I think we are a README file describing what the intention of the library is, why it is needed and what it solves.

File: Tests/HelpersTests/HelpersTests.swift (hunk 1 above)
Comment (Emadgnia) [reply Tests/HelpersTests/HelpersTests.swift:32]: Because I moved the extension from Version to Helper module and this is more of the extension test than a logic for version test I believe!

File: Tests/MaintenanceTests/MaintenanceTests.swift (hunk 2 above)
Comment (Emadgnia) [reply Tests/MaintenanceTests/MaintenanceTests.swift:49]: How would you think that they can be merged?

File: Tests/MaintenanceTests/MaintenanceTests.swift (hunk 3 above)
Comment (Emadgnia) [reply Tests/MaintenanceTests/MaintenanceTests.swift:11]: On those tests when one of test one of them the other one is false, how would you suggest me to test the false?
like false false or what?

File: Sources/Maintenance/Maintenance.swift (hunk 5 above)
Comment (Emadgnia) [reply Sources/Maintenance/Maintenance.swift:19]: Yes, will add in the next commit

File: Sources/Maintenance/Maintenance.swift (hunk 6 above)
Comment (Emadgnia) [reply Sources/Maintenance/Maintenance.swift:52]: will add in next commit

File: Sources/Maintenance/Maintenance.swift (hunk 7 above)
Comment (Emadgnia) [reply Sources/Maintenance/Maintenance.swift:111]: My understanding is fullscreen has higher priority, we don't care if both are true

File: Sources/Maintenance/Models/StatusType.swift (hunk 13 above)
Comment (Emadgnia) [reply Sources/Maintenance/Models/StatusType.swift:9]: Yes, next commit

File: Sources/VersionCheck/VersionCheck.swift (hunk 14 above)
Comment (Emadgnia) [reply Sources/VersionCheck/VersionCheck.swift:12]: Same answer

File: Sources/Helpers/CustomError.swift (hunk 10 above)
Comment (Emadgnia) [reply Sources/Helpers/CustomError.swift:9]: Yes, next commit

File: Sources/Helpers/RemoteConfigHelper.swift (hunk 11 above)
Comment (Emadgnia) [reply Sources/Helpers/RemoteConfigHelper.swift:10]: They need to know the type of error that they get, also it's a helper if you want you can use it

Comment (Emadgnia) [reply Sources/Helpers/RemoteConfigHelper.swift:10]: and then I add docs in next commit

File: Tests/MaintenanceTests/MaintenanceTests.swift (hunk 2 above)
Comment (patrickusiewicz) [reply Tests/MaintenanceTests/MaintenanceTests.swift:49]:
```suggestion
    func testFullscreen() async throws {
        let input = Maintenance.isThereAnyMaintenance(
            fullscreen: true,
            banner: false,
            fullscreenDetail: Maintenance.Detail(title: "FullScreen", body: "Body Test"),
            bannerDetail: Maintenance.Detail(title: "Banner", body: "Body Test"))
        
        XCTAssertEqual(input?.0, Maintenance.StatusType.fullscreen, "The output is not expected!")
        XCTAssertEqual(input?.1, Maintenance.Detail(title: "FullScreen", body: "Body Test"), "The output is not expected!")
    }
    func testBanner() async throws {
        let input = Maintenance.isThereAnyMaintenance(
            fullscreen: false,
            banner: true,
            fullscreenDetail: Maintenance.Detail(title: "FullScreen", body: "Body Test"),
            bannerDetail: Maintenance.Detail(title: "Banner", body: "Body Test"))
            
        XCTAssertEqual(input?.0, Maintenance.StatusType.banner, "The output is not expected!")
        XCTAssertEqual(input?.1, Maintenance.Detail(title: "Banner", body: "Body Test"), "The output is not expected!")
    }
```

File: Tests/MaintenanceTests/MaintenanceTests.swift (hunk 3 above)
Comment (patrickusiewicz) [reply Tests/MaintenanceTests/MaintenanceTests.swift:11]: Something like:
```
Maintenance.isThereAnyMaintenance(
            fullscreen: false,
            banner: false,
            fullscreenDetail: Maintenance.Detail(title: "", body: ""),
            bannerDetail: Maintenance.Detail(title: "", body: ""))
```
?

File: Sources/Maintenance/Maintenance.swift [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+            bannerDetail: bannerDetail)
+        
+        
+    }
+    
+    static func isThereAnyMaintenance(fullscreen: Bool,
+                                      banner: Bool,
+                                      fullscreenDetail: Detail,
+                                      bannerDetail: Detail) -> (StatusType, Detail)? {
+
```
Comment (Nicolaidam) [inline Sources/Maintenance/Maintenance.swift:110]: For me its very hard to read the return type of this function, I think this more readable so we have the -> on its own line so you dont miss the return type. I was thinking this was a void since i didnt see the ->.

I think our linting tool also prefers this way right?

    static func isThereAnyMaintenance(
        fullscreen: Bool,
        banner: Bool,
        fullscreenDetail: Detail,
        bannerDetail: Detail
    ) -> (StatusType, Detail)? {
        
        if fullscreen {
            return (.fullscreen, fullscreenDetail)
        }
        
        if banner {
            return (.banner, bannerDetail)
        }
        
        return nil
    }

File: Sources/Maintenance/Models/Detail.swift [hunk 16]
```diff
@@ -0,0 +1,14 @@
+//
+//  File.swift
+//  
+//
+//  Created by Emad Ghorbania on 14/12/2022.
+//
+
+import Foundation
```
Comment (Nicolaidam) [inline Sources/Maintenance/Models/Detail.swift:8]: You can remove foundation here

File: Sources/Maintenance/Models/StatusType.swift [hunk 17]
```diff
@@ -0,0 +1,14 @@
+//
+//  File.swift
+//  
+//
+//  Created by Emad Ghorbania on 14/12/2022.
+//
+
+import Foundation
```
Comment (Nicolaidam) [inline Sources/Maintenance/Models/StatusType.swift:8]: You can remove foundation here

**Review by Nicolaidam — COMMENTED** [review]
Good work, i added a few comments 😎

**Review by Nicolaidam — APPROVED** [review]
